[Go] New Concept Exercise: conditionals

andres-zartab opened the pull request:
Hopefully fixes #428.
First attempt at creating this concept exercise. It is still a WIP but any input or suggestion is welcome 😄

tehsphinx commented on languages/go/exercises/concept/conditionals/.meta/example.go:
```diff
@@ -0,0 +1,81 @@
+	default:
+		return "H"
+	}
+}
```
This would be nice for a practice exercise but is too complex for a concept exercise. Maybe something where e.g. 2 ifs are needed, one with an `||`, one with an `&&`? Could also be 2 small functions instead.

tehsphinx commented on languages/go/exercises/concept/conditionals/.meta/example.go:
```diff
@@ -0,0 +1,81 @@
+		return true
+	}
+	return false
+}
```
I'd use this function or the `Blackjack` function, not both as it is the same.

tehsphinx commented on languages/go/exercises/concept/conditionals/.docs/instructions.md:
```diff
@@ -0,0 +1,32 @@
+
+This strategy is as follows:
+
+- If you have a Blackjack and the dealer does not have an ace, a figure or a ten then you automatically win. If the dealer does have any of those cards then you'll have to stand and wait for the reveal of the other card.
```
We should explain what a `Blackjack` is...

tehsphinx commented on languages/go/exercises/concept/conditionals/.docs/instructions.md:
```diff
@@ -0,0 +1,32 @@
+# Instructions
```
A general comment on the `instructions`: Have a look at the instructions of other existing concept exercises (e.g. in C# folder). A story is very good and then it is good to go into a small section explaining each function.

tehsphinx commented:
Note: If you want to keep the story, there is always the possibility to have existing code in the stub the student doesn't have to write themselves. It should be in a way though so the students does not need to understand that code and it is still clear what needs to be done.

andres-zartab commented:
Awesome feedback @tehsphinx, will address these issues and try to complete the exercise shortly :)

andres-zartab commented:
Restructured the exercise so it wouldn't contain too many cases. Also added introduction.md and updated the instructions.md file (hints.md and after.md still missing). @tehsphinx please let me know what you think :)

tehsphinx commented on languages/go/exercises/concept/conditionals/.docs/instructions.md:
```diff
@@ -0,0 +1,64 @@
+
+### 2. Determine if two cards make up a Blackjack.
+
+Implement a function that returns true if two cards are form a Blackjack, false otherwise.
```
```suggestion
Implement a function that returns true if two cards form a Blackjack, false otherwise.
```

andres-zartab replied:
corrected

tehsphinx commented on languages/go/exercises/concept/conditionals/.docs/instructions.md:
```diff
@@ -0,0 +1,64 @@
+
+Although not optimal yet, you will follow the strategy your friend Alex has been developing, which is as follows:
+
+- If you have a Blackjack (an ace and any card worth 10 points) and the dealer does not have an ace, a figure or a ten then you automatically win. If the dealer does have any of those cards then you'll have to stand and wait for the reveal of the other card.
```
> an ace and any card worth 10 points

Maybe replace with: `2 cards with a value of 21`

andres-zartab replied:
corrected

tehsphinx commented on languages/go/exercises/concept/conditionals/.docs/instructions.md:
````diff
@@ -0,0 +1,64 @@
+Implement a function that returns the string representation of a decision given your cards, the dealer's card and Alex's strategy.
+
+```go
+choice := FirstTurn("ace", "ace", "two")
````
```suggestion
playerCard1 := "ace"
playerCard2 := "ace"
dealerCard := "two"
choice := FirstTurn(playerCard1, playerCard2, dealerCard)
```

Probably easier to understand which is which.

andres-zartab replied:
much clearer this way

tehsphinx commented on languages/go/exercises/concept/conditionals/.meta/example.go:
```diff
@@ -0,0 +1,60 @@
+	default:
+		return ""
+	}
+}
```
Wondering if this still gives the student too much freedom to do unidiomatic things which would be hard to check automatically. (Since we want 95% solutions of the concept exercises be handled automatically.)

My suggestion:

Split this in more functions:
```go
func FirstTurn(card1, card2, dealerCard string) string {
	handScore := ParseCard(card1) + ParseCard(card2)
	dealerScore := ParseCard(dealerCard)
	isBlackjack := Blackjack(card1, card2)

	s := largeHand(handScore, isBlackjack, dealerScore)
	if s != "" {
		return s
	}
	return smallHand(handScore, dealerScore)
}

func largeHand(handScore int, isBlackjack bool, dealerScore int) string {
	// Single if statement
	if handScore == 22 {
		return "P"
	}
	// Nested if statement (no else)
	if isBlackjack {
		if dealerScore < 10 {
			return "W"
		}
		return "S"
	}
	return ""
}

func smallHand(handScore int, dealerScore int) string {
	// Switch statement
	switch {
	case (handScore >= 17) || (handScore >= 12 && dealerScore <= 6):
		return "S"
	case (handScore <= 11) || (handScore >= 12 && dealerScore > 6):
		return "H"
	default:
		return ""
	}
}
```

`FirstTurn` would be part of the stub and the student would not have to implement it. The other 2 functions would need to be implemented by the student.

What do you think? @andres-zartab 

@ErikSchierboom: What do you think? Or is this ok as it is?

ErikSchierboom replied:
I am wondering if we could somehow simplify things a bit. There is quite a bit of code involved, and I'm wondering if this would lead to many different representations (which would make commenting on those representations less efficient). I like the `ParseCard` and `Blackjack` functions. Maybe we could simplify or replace the `FirstTurn` function a bit?

tehsphinx replied:
I really like the `smallHand` function I mentioned above as a function the student could implement.

andres-zartab replied:
Will give some thought to this now and propose something shortly

ErikSchierboom commented on languages/go/exercises/concept/conditionals/.docs/after.md:
```diff
@@ -0,0 +1 @@
+# After
```
If the document is empty, I'd remove it.

andres-zartab replied:
updated!

ErikSchierboom commented on languages/go/exercises/concept/conditionals/.docs/hints.md:
```diff
@@ -0,0 +1,3 @@
+# Hints
```
If the document is empty, I'd remove it.

andres-zartab replied:
updated!

ErikSchierboom commented on languages/go/exercises/concept/conditionals/.docs/introduction.md:
```diff
@@ -0,0 +1,8 @@
+# Introduction
+
+### Conditionals
```
I'd remove this heading.

```suggestion
```

andres-zartab replied:
done

ErikSchierboom commented on languages/go/exercises/concept/conditionals/.docs/introduction.md:
```diff
@@ -0,0 +1,8 @@
+
+### Conditionals
+
+Conditionals in Go are similar to conditionals in other languages. The underlying type of any conditional operation is the `bool` type, which can have the value of `true` or `false`. Conditionals are often used as flow control mechanisms to check for various conditions. For checking a particular case an `if` statement can be used, which executes its code if the underlying condition is `true`. For ...
```
I really like this description!

andres-zartab replied:
thanks! :)

ErikSchierboom commented on languages/go/exercises/concept/conditionals/.docs/introduction.md:
```diff
@@ -0,0 +1,8 @@
+
+Conditionals in Go are similar to conditionals in other languages. The underlying type of any conditional operation is the `bool` type, which can have the value of `true` or `false`. Conditionals are often used as flow control mechanisms to check for various conditions. For checking a particular case an `if` statement can be used, which executes its code if the underlying condition is `true`. For ...
+
+To learn more about this topic it is recommended to check these sources:
```
These additional resources should probably be part of the `hints.md` document instead.

andres-zartab replied:
updated!

ErikSchierboom commented on languages/go/exercises/concept/conditionals/conditionals.go:
```diff
@@ -0,0 +1,16 @@
+
+// ParseCard returns the integer value of a card following blackjack ruleset.
+func ParseCard(card string) int {
+	panic("not implemented")
```
```suggestion
	panic("Please implement the ParseCard function")
```

ErikSchierboom commented on languages/go/exercises/concept/conditionals/conditionals.go:
```diff
@@ -0,0 +1,16 @@
+
+// Blackjack returns true if the player has a blackjack, false otherwise.
+func Blackjack(card1, card2 string) bool {
+	panic("not implemented")
```
```suggestion
	panic("Please implement the Blackjack function")
```

ErikSchierboom commented on languages/go/exercises/concept/conditionals/conditionals.go:
```diff
@@ -0,0 +1,16 @@
+
+// FirstTurn returns the semi-optimal decision for the first turn, given the cards of the player and the dealer.
+func FirstTurn(card1, card2, dealerCard string) string {
+	panic("not implemented")
```
```suggestion
	panic("Please implement the FirstTurn function")
```

ErikSchierboom commented on languages/go/exercises/concept/conditionals/.meta/design.md:
```diff
@@ -0,0 +1,34 @@
+## Concepts
+
+- `boolean type`
+- `if`
```
```suggestion
- `conditionals-if`
```

ErikSchierboom commented on languages/go/exercises/concept/conditionals/.meta/design.md:
```diff
@@ -0,0 +1,34 @@
+
+- `boolean type`
+- `if`
+- `switch`
```
```suggestion
- `conditionals-switch`
```

ErikSchierboom commented on languages/go/exercises/concept/conditionals/.meta/design.md:
```diff
@@ -0,0 +1,34 @@
+
+## Concepts
+
+- `boolean type`
```
```suggestion
- `booleans`
```

ErikSchierboom commented on languages/go/exercises/concept/conditionals/.meta/design.md:
```diff
@@ -0,0 +1,34 @@
+- iterations
+- continue (as there are no iterations)
+
+## Concepts
```
I'm suggesting a slightly normalized version of the concepts used below, to make things more consistent cross-tracks.

andres-zartab replied:
corrected

andres-zartab commented:
Hi @tehsphinx and @ErikSchierboom! Thanks for all your feedback and sorry for being away so long. Now I have more free time so I will check and try to resolve everything :)

ErikSchierboom commented:
@andres-zartab Sure, great that you're back! In the meantime, there has been another boolean conditions exercise submitted to the JavaScript track: https://github.com/exercism/v3/pull/841 I don't know if it can help you, but it might be interesting to look at that PR to see how they went about designing the exercise.
